amethyst_renderer/src/cam.rs

```diff
@@ -53,25 +53,128 @@ impl From<Projection> for Matrix4<f32> {
 }
 
 /// Camera struct.
-///
-/// TODO: Add more convenience methods, refine API.
 #[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
 pub struct Camera {
+    /// Initial forward vector of the camera.
+    pub base_forward: Vector3<f32>,
+    /// Calculated center point of the camera.
+    center: Vector3<f32>,
+    /// Trigger re-calculations if values were changed.
+    pub dirty: bool,
     /// Location of the camera in three-dimensional space.
     pub eye: Point3<f32>,
     /// Graphical projection of the camera.
     pub proj: Matrix4<f32>,
-    /// Forward vector of the camera.
-    pub forward: Vector3<f32>,
-    /// Right vector of the camera.
-    pub right: Vector3<f32>,
+    /// Rotation of the camera in three-dimensional space.
+    pub rotation: Quaternion<f32>,
     /// Upward elevation vector of the camera.
     pub up: Vector3<f32>,
 }
 
 impl Camera {
+    /// Create a new camera from the given values.
+    pub fn new(
+        projection: Matrix4<f32>,
+        position: Point3<f32>,
+        look_at: Point3<f32>,
+        up: Vector3<f32>
+    ) -> Self {
+        if position == look_at {
+            panic!("Position and LookAt cannot have the same coordinates!");
+        }
+        
+        let base_forward = (look_at - position);
+        Self {
+            base_forward: base_forward,
+            center: (position + base_forward).to_vec(),
+            dirty: false,
+            eye: position,
+            proj: projection,
+            rotation: Quaternion::look_at(base_forward, up).normalize(),
+            up: up,
+        }
+    }
+    
+    /// Move the camera relatively to its current position, but independently from its orientation.
+    pub fn move_absolute(&mut self, direction: Vector3<f32>) -> &mut Self {
```

Comment:
Maybe rename `move_absolute` to `move_global` and `move_relative` to `move_local`? Because `move_absolute` would suggest setting the position of the camera and `move_relative` would suggest shifting the position of the camera relative to it's initial position by adding a `direction` vector to it. This unwarranted suggestion would add unnecessary cognitive load on API users. Same change would apply to all of the other `*_absolute`/`*_relative` methods below for consistency if you will decide to make this change.

Also, I think it would be better to change `direction` to `shift` or `translation` because `direction` would suggest a unit vector, which only defines direction, when in fact it is an arbitrary not-necessarily-unit vector representing a shift either in the global coordinate system or in the local coordinate system of the camera.

And it would be nice to add more detailed doc comments for these methods because they wouldn't be at all trivial to a person learning to use this API for the first time. Maybe something like:
```rust
/// Move the camera by `shift` vector in the global coordinate system,
/// the resulting move doesn't depend on camera's orientation.
pub move_global(...) {...}
/// Move the camera by `shift`vector in the local coordinate system, i.e. if `shift == Vec3(dx, dy, dz)`
/// the camera will move by `dx` in the `forward` direction, by `dy` in the `right` direction and,
/// `dz` in the `up` direction. Of course this means that the resulting move depends on the orientation of
/// this `forward`, `right`, `up` triplet of vectors.
pub fn move_local(...) { ... }
```

Reply:
Both hands for those names. "Global" and "local" are widely used to describe basis. In 3d modeling software for example.

Reply:
I agree, I will change the names of the methods and parameters.